### PR TITLE
[Enhancement] Async tasks should work 100%

### PR DESCRIPTION
**Pull request by tom-pytel**
These changes should address all remaining async-related problems not handled in #88. The `spans` list has been moved from an instance variable in `SpanContext` to a global task-local variable which allows "forking" it for new subtasks so that multiple concurrent async subtasks don't interfere with one another. `spans` is now a top-level module var because Python stipulates that `ContextVar` should be such and only one context exists at any given time so it is essentially a singleton anyway.

### CODE REVIEW

File: skywalking/trace/context.py [hunk 1]
```diff
@@ -27,9 +25,52 @@
 from skywalking.utils.counter import Counter
 
 
+try:  # attempt to use async-local instead of thread-local context and spans
+    import contextvars
+
+    __local = contextvars.ContextVar('local')
+    __spans = contextvars.ContextVar('spans')  # this needs to be a per-task variable, can't be part of __local
+    _spans = __spans.get
+    _spans_set = __spans.set
```
[the remark this answers is not in the document]
Comment (tom-pytel) [reply skywalking/trace/context.py:34]: Never let a perfectly good piece of code get in the way of a proper linting I guess... Can you make this go away?

[the remark this answers is not in the document]
Comment (TomMD) [reply skywalking/trace/context.py:34]: @tom-pytel Pyre's type warnings are one of the things we think about how to balance the friction and value of in @muse-dev - we're certainly open to any feedback.  Currently the maintainer can click "resolve" on this issue.  Soon (in the next week or two) we'll release a new feature allowing anyone to reply to the bot with "ignore", causing the issue to resolve and adjusting the status bar at the bottom on the PR.

[the remark this answers is not in the document]
Comment (tom-pytel) [reply skywalking/trace/context.py:34]: I am not familiar with the system, is there a comment hint I can include in the source which will make it ignore this particular case?

[the remark this answers is not in the document]
Comment (TomMD) [reply skywalking/trace/context.py:34]: Pyre does have an ignore command (below).  Alternatively, Muse is setup to only surface newly detected bugs so you could just ignore/resolve this comment in via GitHub's button - the note would continue to [exist on the console](https://console.muse.dev/result/apache/skywalking-python/01EQXNB645P9R98PQHJZ1XD7RX) but would not appear as a comment again.  The intent is to minimize any demand for the developer's attention by allowing quick consideration when they arise (and equally quick dismissal if appropriate).

```suggestion
    _spans_set = __spans.set # pyre-ignore
```

**Comment by kezhenxu94** [comment]
> The `spans` list has been moved from an instance variable in `SpanContext` to a global task-local variable which allows "forking" it for new subtasks so that multiple concurrent async subtasks don't interfere with one another.

By "interfere", I'm not perfectly sure what you mean, IMO, subtasks should share the same context as the one where it is spawned, is there another reason to [duplicate the `spans`](https://github.com/apache/skywalking-python/pull/89/files#diff-275344c0ebbcd139d9e6006c72a18a95e05ffeca35f7dfdffc9125fd4621e02fR49-R51).

Also, I didn't succeed to make up a FAILED test case that "a single entry point executing multiple async exit requests simultaneously" (as you mentioned in #88), can you provide a snippet of code so that I can compliment the tests by adding a negative test case.


Other changes look good enough to me, thanks

**Comment by tom-pytel** [comment]
> By "interfere", I'm not perfectly sure what you mean, IMO, subtasks should share the same context as the one where it is spawned, is there other reason to [duplicate the `spans`](https://github.com/apache/skywalking-python/pull/89/files#diff-275344c0ebbcd139d9e6006c72a18a95e05ffeca35f7dfdffc9125fd4621e02fR49-R51)?

They do share the same context but the `spans` list must be different because if not then sibling tasks might wind up with a parent-child relationship since the parent for a new span is the span at the end of the current `spans` list. So for example if you were to have an entry span which then launches two other async spans concurrently then with the previous code the `spans` list might wind up as:
```
[EntrySpan, LocalSpan1, LocalSpan2]
```
With LocalSpan1 as parent to LocalSpan2, even though LocalSpan1 and LocalSpan2 should both have EntrySpan as the parent. Or worse, with ExitSpan the second would overwrite the first if I understand the code correctly and you would wind up with only one ExitSpan where two outgoing requests were made.

With this change each new async task gets its own isolated copy of a `spans` list which will wind up looking like:
```
[EntrySpan, LocalSpan1]
```
for the first sibling task and:
```
[EntrySpan, LocalSpan2]
```
for the second, with correct parents for everyone.

> 
> Also, I didn't succeed to make up a FAILED test case that "a single entry point executing multiple async exit requests simultaneously" (as you mentioned in #88), can you provide a snippet of code so that I can compliment the tests by adding a negative test case.

That kind of thing should no longer fail with this PR (which is the whole point of it) but should fail with #88 and below. If you were not successful in getting a fail case for #88 then that is different and I will look into it tomorrow so do let me know if you are referring to this PR with respect to not being able to make a fail case or #88.

**Comment by kezhenxu94** [comment]
The explanation of "interfere" totally makes sense to me.

> If you were not successful in getting a fail case for #88

That is what I meant, this kind of cases may be hard to reproduce stably but it would be nice if we can add a test case to possibly cover it as our best effort, so I need your help to reproduce it under #88, feel free to do it when you got some time. I'm merging this for now. Thanks again

**Review by kezhenxu94 — APPROVED** [review]

**Comment by tom-pytel** [comment]
> That is what I meant, this kind of cases may be hard to reproduce stably but it would be nice if we can add a test case to possibly cover it as our best effort, so I need your help to reproduce it under #88, feel free to do it when you got some time. I'm merging this for now. Thanks again

```
import time
from skywalking import agent
from skywalking.trace.context import get_context

agent.start()

async def test(num):
    # with get_context().new_local_span(f'child{num}'):
    with get_context().new_exit_span(f'child{num}', '0.0.0.0'):
        from skywalking.trace.context import _spans; print('...', _spans())
        await asyncio.sleep(0.01)  # allow other tasks to tick and start spans
        if num == 1:
            get_context().active_span().raised()  # error a single child

async def main():
    with get_context().new_local_span('parent'):
        await asyncio.gather(
            test(0),
            test(1),
            test(2),
        )

asyncio.run(main())

time.sleep(1)  # allow BG daemon thread to finish updating
```
Compare #88 with this PR.

**Comment by kezhenxu94** [comment]
Thanks.

> P.S. If you have a moment could you have a look at my question #5875 on the main skywalking issues board? I will be moving on to the NodeJS agent so would be nice to be able to test it as I work :)

I will check it soon

**Comment by tom-pytel** [comment]
Out of curiosity, what purpose do Context.capture() and Context.continue() functions serve? Is it to preserve state across possibly asynchronous operation? If so then with this PR they may no longer be necessary.

**Comment by wu-sheng** [comment]
> Out of curiosity, what purpose do Context.capture() and Context.continue() functions serve? Is it to preserve state across possibly asynchronous operation? If so then with this PR they may no longer be necessary.

Let me give an example. Yes, those things designed for across threads, originally from [Java agent core](https://github.com/apache/skywalking/blob/master/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/TracingContext.java#L197-L232). 
The reason to have this is to avoid the lock and get rid of the race condition bug in the tracing process. We are targeting 100% sampling tracing, and analyzing the metrics/topology based on tracing data.

Each language agent has its own choice whether it needs to adopt all modules in the java agent implementation.

**Comment by kezhenxu94** [comment]
> Out of curiosity, what purpose do Context.capture() and Context.continue() functions serve? Is it to preserve state across possibly asynchronous operation?

Yes

> If so then with this PR they may no longer be necessary.

Seems that's the case. Wanna open a pull request?

**Comment by kezhenxu94** [comment]
> Out of curiosity, what purpose do Context.capture() and Context.continue() functions serve? Is it to preserve state across possibly asynchronous operation?

Yes

> If so then with this PR they may no longer be necessary.

Seems like that's the case, feel free to open a pull request

**Comment by tom-pytel** [comment]
I will return to this after the NodeJS agent stuff.

**Pull request merged**